benches/storage_perf.rs

```diff
@@ -0,0 +1,73 @@
+use criterion::{Bencher, Criterion};
+use specs::prelude::*;
+
+#[derive(Clone, Debug)]
+pub struct GridVelocity {
+    pub dx: i16,
+    pub dy: i16,
+}
+impl GridVelocity {
+    pub fn new() -> Self {
+        GridVelocity { dx: 0, dy: 0 }
+    }
+}
+impl Component for GridVelocity {
+    type Storage = VecStorage<Self>;
+}
+
+fn bench_vec(b: &mut Bencher) {
+    let mut vels: Vec<GridVelocity> = (0..1000 * 1000).map(|_i| GridVelocity::new()).collect();
+    b.iter(|| {
+        vels.iter_mut().for_each(|vel| {
+            vel.dx += 1;
+        });
+    });
+}
+
+pub fn bench_vecstorage(b: &mut Bencher) {
+    use specs::storage::UnprotectedStorage;
+    use std::default::Default;
+
+    let mut vel_storage: VecStorage<GridVelocity> = Default::default();
+
+    for i in 0..1000 * 1000 {
+        unsafe {
+            vel_storage.insert(i, GridVelocity::new());
+        }
+    }
+
+    b.iter(|| {
+        for i in 0..1000 * 1000 {
+            let mut vel = unsafe { vel_storage.get_mut(i) };
+            vel.dx += 1;
+        }
+    });
+}
+
+pub fn bench_storage_entity_prefetch(b: &mut Bencher) {
+    let mut world = World::new();
+    world.register::<GridVelocity>();
+    for _i in 0..1000 * 1000 {
+        world.create_entity().with(GridVelocity::new()).build();
+    }
+
+    let mut vel_storage = world.write::<GridVelocity>();
+    let ents = world.entities();
+    let entities: Vec<Entity> = (0..1000 * 1000).map(|i| ents.entity(i)).collect();
+
+    b.iter(|| {
+        for e in entities.iter() {
+            let mut vel = vel_storage.get_mut(*e).unwrap();
+            vel.dx += 1;
+        }
```

Comment:
I am not sure what is the idea behind this prefetch thing. Why don't directly iterate entities? Collecting them to vector is expensive operation which I presume would be performed whenever new things are added to world.

Reply:
I also don't know why are avoiding `join` which is the usual way to use storages. Getting them like that is more expensive as you need to check liveness. (There is some overhead on `join` due to iterating `BitSet`)

Reply:
In my actual use case, I have a different component containing a list of entities, which is built by looking for other entities having this component. So in my case, I can guarantee they will have this component, and I guess I'd handle removal in a separate system running later.

About prefetching: basically I have a Vec<Entity>. I guess I could create a hibitset with that and join it ? otherwise I tried not to bench the entity fetching, given that it is expensive.

Btw joining/iterating on hibitset seems to be crazy expensive too - I merged two systems at some points, because joining twice was slow.

Reply:
> I can guarantee they will have this component

If this is too slow for you and you can guarantee it, then maybe you just want to access the raw storage?